File: apps/user/src/functions/ConnectionGroupHandler.ts
'use strict';


import {UserService} from '../service/UserService';
import {Handler} from "aws-lambda";
import {CreateGroupRequest} from "../model/request/CreateGroupRequest";
import {AddConnectionToGroupRequest} from "../model/request/AddConnectionToGroupRequest";
import {RemoveConnectionFromGroupRequest} from "../model/request/RemoveConnectionFromGroupRequest";
import {SearchGroupsRequest} from "../model/request/SearchGroupsRequest";

const userService = new UserService()

export const createGroup: Handler = async (event) => {
    try {
        console.log("Received request to create connection group")
        const {userId} = event.pathParameters
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)

        const request = new CreateGroupRequest(userId, body.name)
        request.validate()
        const result = await userService.createGroup(request)

        return createResponse(201, result)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const getGroups: Handler = async (event) => {
    try {
        console.log("Received request to get connection groups")
        const {userId} = event.pathParameters
        const result = await userService.getConnectionGroups(userId)

        return createResponse(200, result)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const deleteGroup: Handler = async (event) => {
    try {
        console.log("Received request to delete connection group")
        const {userId, groupId} = event.pathParameters
        const success = await userService.deleteConnectionGroup(userId, groupId)

        return createResponse(success ? 200 : 404, {})
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}


export const exportGroup: Handler = async (event) => {
    try {
        console.log("Received request to export connection group")
        const {userId, groupId} = event.pathParameters
        await userService.exportGroup(userId, groupId)

        return createResponse(200, {})
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const getConnections: Handler = async (event) => {
    try {
        console.log("Received request to get connections in group")
        const {userId, groupId} = event.pathParameters

        const queryParams = (event.queryStringParameters || {})
        const searchParams = typeof queryParams === 'object' ? queryParams : JSON.parse(queryParams)
        const name = searchParams.name
        const limit = searchParams.limit ? parseInt(searchParams.limit) : undefined;
        const skip = searchParams.skip ? parseInt(searchParams.skip) : undefined

        const request = new SearchGroupsRequest(userId, name, limit, skip)
        request.validate()

        const result = await userService.searchGroupConnections(groupId, request)

        return createResponse(200, result)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const addConnection: Handler = async (event) => {
    try {
        console.log("Received request to add connection to connection group")
        const {userId, groupId} = event.pathParameters
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)

        const request = new AddConnectionToGroupRequest(userId, groupId, body.userId)
        request.validate()
        const result = await userService.addConnectionToGroup(request)

        return createResponse(200, result)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const removeConnection: Handler = async (event) => {
    try {
        console.log("Received request to add connection to connection group")
        const {userId, groupId} = event.pathParameters
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)

        const request = new RemoveConnectionFromGroupRequest(userId, groupId, body.userId)
        request.validate()
        const result = await userService.removeConnectionFromGroup(request)

        return createResponse(200, result)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}


function createResponse(statusCode: number, body: object) {
    return {
        statusCode: statusCode || 500,
        body: JSON.stringify(body)
    }
}

File: apps/user/src/functions/connection_group_handler.py
import asyncio
import json
import logging

from shared.authorization.handler_utils import create_response, exception_handler
from src.model.requests import CreateGroupRequest, SearchGroupsRequest, AddConnectionToGroupRequest, \
    RemoveConnectionFromGroupRequest
from src.service.connection_service import ConnectionService

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

connection_service = ConnectionService()


@exception_handler
def create_group(event, context):
    path_params = event.get('pathParameters') or {}
    body = json.loads(event.get('body', '{}'))

    request = CreateGroupRequest(userId=path_params.get("userId"), **body)
    res = connection_service.create_group(request=request)
    return create_response(200, res)


@exception_handler
def get_groups(event, context):
    path_params = event.get('pathParameters') or {}
    res = connection_service.get_connection_groups(user_id=path_params.get("userId"))
    return create_response(200, res)


@exception_handler
def delete_group(event, context):
    path_params = event.get('pathParameters') or {}
    res = connection_service.delete_connection_group(user_id=path_params.get("userId"),
                                                     group_id=path_params.get("groupId"))
    return create_response(200, res)


@exception_handler
def export_group(event, context):
    path_params = event.get('pathParameters') or {}
    res = asyncio.run(connection_service.export_group(user_id=path_params.get("userId"),
                                                      group_id=path_params.get("groupId")))
    return create_response(200, res)


@exception_handler
def get_connections(event, context):
    path_params = event.get('pathParameters') or {}
    query_params = event.get('queryStringParameters') or {}
    request = SearchGroupsRequest(
        userId=path_params.get("userId"),
        name=query_params.get("name"),
        limit=int(query_params.get("limit")) if "limit" in query_params else 25,
        skip=int(query_params.get("skip")) if "skip" in query_params else 0,
    )

    connections = connection_service.search_group_connections(group_id=path_params.get("groupId"),
                                                              request=request)
    return create_response(200, connections)


@exception_handler
def add_connection(event, context):
    path_params = event.get('pathParameters') or {}
    body = json.loads(event.get('body', '{}'))

    request = AddConnectionToGroupRequest(userId=path_params.get("userId"),
                                          groupId=path_params.get("groupId"),
                                          connectedUserId=body.get("userId"))
    res = connection_service.add_connection_to_group(request=request)
    return create_response(200, res)


@exception_handler
def remove_connection(event, context):
    path_params = event.get('pathParameters') or {}
    body = json.loads(event.get('body', '{}'))

    request = RemoveConnectionFromGroupRequest(userId=path_params.get("userId"),
                                               groupId=path_params.get("groupId"),
                                               connectedUserId=body.get("userId"))
    res = connection_service.remove_connection_from_group(request=request)
    return create_response(200, res)
